### DISCUSSION TITLE
User provided device_type of 'cuda', but CUDA is not available. Disabling

### DISCUSSION THREAD

**Discussion by Lohnii** [neonbjb/tortoise-tts#599] [Q&A] [answered]
When trying:
python tortoise/do_tts.py --text "I'm going to speak this" --voice random --preset fast

I get this warning:

Generating autoregressive samples..
C:\Users\User\miniconda3\envs\tortoise\lib\site-packages\torch\amp\autocast_mode.py:204: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')
  
Is there a way to fix this? It makes it take over 40 minutes to generate the samples

I have installed the cuda toolkit (ver 11) and restarted my computern and it's still happening.
My graphics card is an RTX 3060 Ti

This is my 6th try installing tortoise and it's the first time I've gotten this far, I would like to try fixing this without reinstalling

**Comment by Lohnii** [comment, marked as answer]
I was able to fix this by running

conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia

**Reply by Jidis83** [reply]
Man, I think you and I were doing the _exact_ same thing at the same time with the same results. I fought with it the past two nights and must have hit every obstacle imaginable. Each time I got through some error, it would throw a new one at me. I had decided to quit on the next error, but on that final run, there was a tip about something that was on there, but not actually unpacked or installed, so I ran that line and afterward Tortoise magically started to process. I don't even want to think about how much conflicting unnecessary mess is on that system now, but I did it (Lubuntu23) on an external just to try Tortoise.

I did get the same disabled CUDA warning as well, and that was the battle plan for this evening once I leave Windows. I also had run across that same tip about the line which fixed yours, but I believe something in the tail end of it threw an error. I wish I had some idea of what some of these recommended terminal lines were actually doing though and what all the arguments meant. I'll try it again tonight, now that Tortoise actually works.

Take Care (total Linux noob here in case you couldn't tell)

**Reply by Lohnii** [reply]
Basically what i did was go into the pytorch website https://pytorch.org/get-started/locally/ And selected stable, windows, conda, python and cuda 11.7, then i copied the command and pasted it in the command line

**Reply by Jidis83** [reply]
Lohnii,

Thanks! That's the same thing I did to get pytorch, though it was cuda 11.8. I figured that might be involved since most of the instructions online referred to 11.7, so I started trying to remove and reinstall things last night, but screwed the system up so bad it would freeze at some "setvtrgb" service on boot. I think I finally got it back by removing (hopefully) the 470 Nvidea driver for my GT-1030 and letting it put the recommended one back after doing a bunch of other boot problem recommendations. Unfortunately, just about everything but the tortoise-tts folder and Python stuff has been removed now, so I'll have to build it back up tonight. I'm hoping it won't hit me with so many errors, as I've already forgotten what the resolution to each one was.

Take Care

**Reply by Jidis83** [reply]
It's a train wreck here. The re-installs surprised me by working perfectly (with functioning CUDA), and after processing for ages, it throws a CUDA out of memory error. Tried the fixes for that to no avail, then installed a couple CUDA memory optimization packages (no good). I gave up and tried to get the CPU one back up and got two new errors, so I called it quits.

Sorry to hijack this with sob stories, and glad you finally got yours working. I'll see what I can get tonight then figure out who to ask if I still can't get anywhere.

PS- Funny, but I set this Lubuntu system up because it sounded like it was developed on Linux and I'd have less to do to make it work.

**Reply by islewis** [reply]
I was having the same issue, but had already tried to run `conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia`.

The fix had to do with getting the correct `pytorch-cuda` version- I'm running cuda 12.3, and so the version that ended up working for me was pytorch-cuda 12.1:
`conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia`

If anyone else is running into this same issue, try playing around with that version.

**Reply by robkau** [reply]
Thanks @islewis that fixed the issue for me on Windows miniconda environment.  These are the steps I used:  

  

  


Do the Conda installation steps [from here](https://github.com/neonbjb/tortoise-tts#local-installation) and open the Anaconda Powershell Prompt from Start Menu.



Check cuda version is 12.3:
```
nvcc --version

Cuda compilation tools, release 12.3, V12.3.103
```


Reinstall pytorch with your example command:
```
conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia


The following packages will be DOWNGRADED:

  pytorch                                 2.1.1-py3.9_cpu_0 --> 2.1.1-py3.9_cuda12.1_cudnn8_0
  torchaudio                                 2.1.1-py39_cpu --> 2.1.1-py39_cu121
  torchvision                               0.16.1-py39_cpu --> 0.16.1-py39_cu121
```


Now CUDA works! From 3 minute generation to 5 second generation.

```
python .\tortoise\do_tts.py --output_path results --preset ultra_fast --voice geralt --text "time flies like an arrow."

miniconda3\envs\tortoise\lib\site-packages\torch\nn\utils\weight_norm.py:30: UserWarning: torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.
  warnings.warn("torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.")

Generating autoregressive samples..
100%|███████████████████████████████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.55s/it]
Computing best candidates using CLVP
100%|███████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.38it/s]
Transforming autoregressive outputs into audio..
100%|█████████████████████████████████████████████████████████████████████████████████| 30/30 [00:01<00:00, 25.94it/s]
100%|█████████████████████████████████████████████████████████████████████████████████| 30/30 [00:01<00:00, 26.29it/s]
100%|█████████████████████████████████████████████████████████████████████████████████| 30/30 [00:00<00:00, 31.12it/s]

```

**Reply by Digvijay1999** [reply]
My CUDA version is 12.4 and latest CUDA that i can see here [pytorch](https://pytorch.org/get-started/locally/) is 12.1 what to do in this case ?

**Reply by Digvijay1999** [reply]
`conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia`
it worked for me too even with CUDA 12.4
